sound/soc/sof/intel/hda-dsp.c

```diff
@@ -226,10 +226,17 @@ bool hda_dsp_core_is_enabled(struct snd_sof_dev *sdev,
 
 	val = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPCS);
 
-	is_enable = (val & HDA_DSP_ADSPCS_CPA_MASK(core_mask)) &&
-		    (val & HDA_DSP_ADSPCS_SPA_MASK(core_mask)) &&
-		    !(val & HDA_DSP_ADSPCS_CRST_MASK(core_mask)) &&
-		    !(val & HDA_DSP_ADSPCS_CSTALL_MASK(core_mask));
+#define MASK_IS_EQUAL(v, m, field) ({	\
+	uint32_t _m = field(m);		\
+	((v) & _m) == _m;		\
+})
+
+	is_enable = MASK_IS_EQUAL(val, core_mask, HDA_DSP_ADSPCS_CPA_MASK) &&
+		MASK_IS_EQUAL(val, core_mask, HDA_DSP_ADSPCS_SPA_MASK) &&
+		!(val & HDA_DSP_ADSPCS_CRST_MASK(core_mask)) &&
+		!(val & HDA_DSP_ADSPCS_CSTALL_MASK(core_mask));
```

Comment:
I don't get why we need to test if all bits are equal for SPA and CPA, but not for CRST and CSTALL?

Why not use MASK_EQUAL everywhere?

Reply:
@plbossart because a 0 means all bits are equal to 0 :-)

Reply:
not following, sorry

Reply:
@plbossart the original code was
```
is_enable = (val & mask1) && (val & mask2) && !(val & mask3) && !(val & mask4);
```
so, if, say, one out of 2 cores were up, the first two conditions would produce `true` but would actually be wrong. They had to be `((val & mask1) == mask1) && ((val & mask2) == mask2)` to make sure that all cores were running. Whereas conditions 3 and 4 are checking that no cores, that should be running are in reset or in stalled state. So they're each checking `((val & maskX) == 0)` and that is already correct, that is already checking that none of the requested bits are set.

Reply:
ok. now I get it.

Should there be a Fixes tag if the commit subject says 'fix'?

Reply:
@plbossart this code is there from the original file commit in 2019 747503b1813a3 ("ASoC: SOF: Intel: Add Intel specific HDA DSP HW operations")

Reply:
right, so can you add this to the commit message to that this fix gets applied to stable versions?

Reply:
@plbossart updated, although I'm not sure stable kernels need it. I *think* it works as is correctly too, because if none of the indicated cores are in reset or stalled, they *probably* are running. So, in practice I think this test delivers correct results now as is too. I think this is more of a correctness fix. But in theory, if it's possible, that a core isn't in reset and isn't stalled and still it isn't running, then this fix will catch that.